### DISCUSSION TITLE
Need some help with my CDC+HID config

### DISCUSSION THREAD

**Discussion by TT-392** [hathach/tinyusb#1927] [Q&A] [answered]
From my very very limited understanding of this stuff, USB keyboard HID data is 8 bytes, the first byte of which is the modifier byte, then a byte that always seems to be 0, maybe this byte means something, then the remaining bytes are the currently pressed keys.

This is also what I am seeing from keyboards I have laying around in usbmon:
`jffff9d598fcf9500 1407593751 C Ii:3:056:1 0:8 8 = 02000f0e 0d000000
`
Though when I modify the hid_composite example, and have it send shift+a, I am getting:
`ffff9d598e300780 1991566782 C Ii:3:063:1 0:4 9 = 01020004 00000000 00
`
I guess that first byte  is some sort of ID (from the comment in tusb_config.h in the examples), And then the 8 bytes I am used to follow. This seems functional.

But in my actual project, I have a setup where I am using both CDC and HID, and with that project's config files, I am getting something that looks like the following:
`ffff9d598e300840 2556380148 C Ii:3:064:4 0:8 8 = 02020004 00000000`
This looks like the result I was getting in the modified hid_composite example, except that the last byte is missing. The pc seems to interpret the first byte as a modifier byte, and the second byte (the actual modifier byte) is ignored, so I guess it just always treats the last 8 bytes as the keyboard HID data?

I am assuming the fact that I am missing that byte has something to do with the fact that my `CFG_TUD_HID_EP_BUFSIZE` in my tusb_config.h being set to 8 instead of 16, like in the example.
but if I change that value, for some reason, after I press one key, I just seem  get a constant stream of nonsensical looking usb data. My best guess is that this is caused because, somewhere, something in the config is set incorrectly, causing something to go wrong deeper in the code when I increase the `CFG_TUD_HID_EP_BUFSIZE`.

I am quite lost am a bit lost on how to fix that. Is there maybe something that has to be changed in the usb_descriptor to be able to send more data without problems or something like that?

Here are my usb_descriptors and usb_config.h if that helps
https://github.com/TT-392/arisu-split-keyboard-firmware/tree/master/code/src/tinyusb

edit:
I should probably add that my config files come from an older example file I found, which has HID+CDC+WEBUSB, where I ripped out the WEBUSB stuff, since I wasn't able to get stuff to work when trying to combine the HID and CDC examples, but I was able to rip out the WEBUSB stuff (after a lot of trying). (Honestly, I don't really know what I am doing)

**Comment by hathach** [comment, marked as answer]
for composite report (multiple device for 1 HID interface), first byte is report id, the rest is the same. e.g keyboard in composite will have 1 report id + 8 bytes payload = 9 bytes.

**Reply by TT-392** [reply]
I didn't realize that the hid report would be interperated differently after removing the other hid endpoint (is what I guess they are called) from the example. Also, I didn't realize that `tud_hid_keyboard_report(uint8_t report_id, uint8_t modifier, uint8_t keycode[6]);` would try to send that extra byte if the report_id was set to something other than zero (I just had it set to my ITF_NUM_KEYBOARD, causing something to get seriously messed up).

So ye, I just had to replace report_id in my tud_hid_keyboard_report call with a 0, that fixed it, tnx.
